### PR TITLE
Implement journalling for state objects

### PR DESCRIPTION
**Pull request by Arachnid**
This PR implements a write journal that records all changes to state objects, and uses it to rewind state changes in the event of a rollback. This enables us to avoid the need to copy entire state objects when a call is made.

### CODE REVIEW

File: core/state/state_object.go [hunk 1]
```diff
@@ -280,6 +298,16 @@ func (self *StateObject) Code(db trie.Database) []byte {
 }
 
 func (self *StateObject) SetCode(codeHash common.Hash, code []byte) {
+	prevcode := self.Code(self.db.db)
+	self.db.journal = append(self.db.journal, codeChange{
+		account:  &self.address,
+		prevhash: self.CodeHash(),
```
Comment (holiman) [inline core/state/state_object.go]: Won't prevhash in all cases be the 0-hash? Or is there some hidden complexity here when suicides happen?

**Comment by trustfarm-dev** [comment]
here's @Arachnid 's commits.
followings are 
https://github.com/trustfarm-dev/geth-1.4.15-arachnid-patch     for ubunut builds.

it's run well, low CPU consumption and now catch-up block height .
I'm running parity node  / geth node same machine and different machine. 4 nodes I'm operating.

eth.trustfarm.io   pool 's block height's go well.

**Comment by trustfarm-dev** [comment]
@Arachnid Ok. Good consensus, but In real fields, I can't endure stop geth nodes.
above contract is will consensus in near future.
If that has completed then, I'll updates again. For Never Stop Nodes!
Thanks. for your contributions.

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: 57208ad403148a5b22aef99c2db2cdf3fb330599
  * Commit subjects should be kept under 80 characters

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Comment by trustfarm-dev** [comment]
@Arachnid , https://github.com/trustfarm-dev/geth-1.4.15-arachnid-patch/blob/master/geth-v1.4.15-arachnid-OOG 
has applied build of your contract OOG Test builds.
What Can Help you? or Do want Any logs?

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: a604209662e15eb3d7dda066888b779138212a8c
  * Commit subjects should be kept under 80 characters

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Comment by trustfarm-dev** [comment]
@Arachnid 
Here's journal , setBalance patch builds.
https://github.com/trustfarm-dev/geth-1.4.15-arachnid-patch/blob/master/geth-v1.4.15-arachnid-OOG-jrn

**Comment by trustfarm-dev** [comment]
@Arachnid  my public pool has blocks! with your patched geth code. mined uncles Luck with very low hash 500MH.

https://etherscan.io/uncle/0x4ab6e5d0b32aee35d4f1aac05bac3d58a80c28b76023968002f2d6fc2f988fb2

**Comment by fjl** [comment]
See #3092, which contains the squashed and improved version.

**Pull request closed without merging**